experimental/harfbuzz/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+    HB_UNICODE_GENERAL_CATEGORY_CURRENCY_SYMBOL as u8, // CurrencySymbol = 25,
+    HB_UNICODE_GENERAL_CATEGORY_MODIFIER_SYMBOL as u8, // ModifierSymbol = 26,
+    HB_UNICODE_GENERAL_CATEGORY_OTHER_SYMBOL as u8, // OtherSymbol = 27,
+    HB_UNICODE_GENERAL_CATEGORY_INITIAL_PUNCTUATION as u8, // InitialPunctuation = 28,
+    HB_UNICODE_GENERAL_CATEGORY_FINAL_PUNCTUATION as u8, // FinalPunctuation = 29,
+];
+
+unsafe extern "C" fn icu4x_hb_unicode_combining_class(
+    _ufuncs: *mut hb_unicode_funcs_t,
+    unicode: hb_codepoint_t,
+    user_data: *mut c_void,
+) -> hb_unicode_combining_class_t {
+    (*(user_data as *mut CanonicalCombiningClassMap))
+        .get32(unicode)
+        .0 as hb_unicode_combining_class_t
+}
+
+unsafe extern "C" fn icu4x_hb_unicode_combining_class_destroy(user_data: *mut c_void) {
+    let _ = Box::from_raw(user_data as *mut CanonicalCombiningClassMap);
+}
+
+#[allow(clippy::indexing_slicing)]
+unsafe extern "C" fn icu4x_hb_unicode_general_category(
+    _ufuncs: *mut hb_unicode_funcs_t,
+    unicode: hb_codepoint_t,
+    user_data: *mut c_void,
+) -> hb_unicode_general_category_t {
+    // Indexing is OK, because `GeneralCategory` data is validated upon
+    // deserialization so that there can be no out-of-range `GeneralCategory`
+    // values (which would be UB to materialize). `GeneralCategory` is a
+    // stable exhaustive enum, and the length of `ICU4X_GENERAL_CATEGORY_TO_HARFBUZZ`
+    // matches the number of enum items, so the index will always be in range here.
+    ICU4X_GENERAL_CATEGORY_TO_HARFBUZZ[(*(user_data as *mut CodePointMapData<GeneralCategory>))
+        .as_borrowed()
+        .get32(unicode) as usize] as hb_unicode_general_category_t
+}
+
+unsafe extern "C" fn icu4x_hb_unicode_general_category_destroy(user_data: *mut c_void) {
+    let _ = Box::from_raw(user_data as *mut CodePointMapData<GeneralCategory>);
+}
+
+unsafe extern "C" fn icu4x_hb_unicode_compose(
+    _ufuncs: *mut hb_unicode_funcs_t,
+    a: hb_codepoint_t,
+    b: hb_codepoint_t,
+    ab: *mut hb_codepoint_t,
+    user_data: *mut c_void,
+) -> hb_bool_t {
+    // Could we trust HarfBuzz to pass valid scalar values?
+    // `hb_buffer_add_codepoints` seems to allow the introduction
+    // of non-scalar-value items to the HarfBuzz buffer. Should we
+    // treat that as conceptually `unsafe` so that if the app has
+    // introduced values that are not valid scalar values, we'd
+    // be OK with UB? OTOH, HarfBuzz docs say "code points", so
+    // following docs to the letter could still result in
+    // surrogate code points even if they don't make sense.
+    let first = if let Some(first) = core::char::from_u32(a) {
+        first
+    } else {
+        return false as hb_bool_t;
+    };
+    let second = if let Some(second) = core::char::from_u32(b) {
+        second
+    } else {
+        return false as hb_bool_t;
+    };
+    if let Some(c) = (*(user_data as *mut CanonicalComposition)).compose(first, second) {
```

Comment:
Thought: @echeran explained to me how the Hangul composition works, and how it is a formula that requires the input `char` to be valid; however, I wonder whether it is possible to make the ICU4X `compose` function take `u32` instead of `char` and GIGO inside of ICU4X, to make this call site a bit simpler.

Reply:
The limiting factor isn't Hangul but `Char16TrieIterator::next()` taking a `char`.